docs/dev/content/implementation.rst

```diff
@@ -0,0 +1,116 @@
+``ContentDatabaseModel`` is used as a marker so that the content_db_router knows to query against the content database only if the model inherits the ContentDatabaseModel.
+
+The tree structure is established by the `parent` field that is a foreign key pointing to another ContentNode ojbect. You can also create a symmetric relationship using the `related` field, or an asymmetric field using the `is_prerequisite` field.
+
+File
+----
+
+The ``File`` model also inherits from ``ContentDatabaseModel``.
+
+To find where the source file is located, the class method ``get_url`` uses the ``checksum`` filed and ``settings.CONTENT_SOURCE_DIR`` to calculate the file path. Every source file is named based on its MD5 hash value(this value is also stored in the ``checksum`` field) and stored in namespaced folder under the directory specified in ``settings.CONTENT_SOURCE_DIR``. Because it's likely to have thousands of source files, some file system cannot handle flat folder with large number of files very well, we create namespaced subfolder to imporve the performance. So the eventual file path can look something like:
+
+    ``user/desktop/my_learning_material/98/9808fa7c560b9801acccf0f6cf74c3ea.mp4``
+
+As you can see, it is fine to store your source files outside of the kolibri project folder as long as you set the ``settings.CONTENT_SOURCE_DIR`` accordingly.
+
+The front-end will then use the ``extension`` field to decide which content player should be used. When the ``supplementary`` field's value is ``True``, that means this File object isn't necessary and can display the content without it. For example, we will mark caption(subtitle) file as supplementary.
+
+Content Constants
+-----------------
+
+A Python module that stores constants for the ``kind`` field in ContentNode model and the ``preset`` field and ``extension`` field in File model.
+
+.. automodule:: kolibri.content.constants.content_kinds
+.. automodule:: kolibri.content.constants.extensions
+.. automodule:: kolibri.content.constants.presets
+
+Workflows
+---------
+
+There are two workflows we currently designed to handle content UI rendering and content playback rendering
+
+- Content UI Rendering
+
+1. Start with a ContentNode object.
+2. Get the associated File object that has the ``thumbnail`` field being True.
+3. Get the thumbnail image by calling this File's ``get_url`` method.
+4. Determine the template using the ``kind`` field of this ContentNode object.
+5. Renders the template with the thumbnail image.
+
+
+- Content Playback Rendering
+
+1. Start with a ContentNode object.
+2. Select a preset constant based on this object's ``kind`` field.
+3. Use the preset constant to retrieve the queryset of all the File objects that have foreign key to this ContentNode object.
+4. Use the ``thumbnail`` field as a filter on this queryset to get the File object and use this File object's ``checksum`` field to get the source file(the thumbnail image)
```

Comment:
Rather than mention checking `checksum`, probably clearer to refer to `get_url` like you do in item 5 and elsewhere.